execution/exchange/dedup.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package exchange
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/thanos-community/promql-engine/execution/model"
+)
+
+type dedupSample struct {
+	t int64
+	v float64
+}
+
+// The dedupCache is an internal cache used to deduplicate samples inside a single step vector.
+type dedupCache []dedupSample
+
+type dedupOperator struct {
+	once   sync.Once
+	series []labels.Labels
+
+	pool *model.VectorPool
+	next model.VectorOperator
+	// outputIndex is a slice that is used as an index from input sample ID to output sample ID.
+	outputIndex []uint64
+	outputCache dedupCache
+}
+
+func NewDedupOperator(pool *model.VectorPool, next model.VectorOperator) model.VectorOperator {
+	return &dedupOperator{
+		next: next,
+		pool: pool,
+	}
+}
+
+func (d *dedupOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	var err error
+	d.once.Do(func() { err = d.loadSeries(ctx) })
+	if err != nil {
+		return nil, err
+	}
+
+	in, err := d.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		return nil, nil
+	}
+
+	result := d.pool.GetVectorBatch()
+	for _, vector := range in {
+		for i, inputSampleID := range vector.SampleIDs {
+			outputSampleID := d.outputIndex[inputSampleID]
+			d.outputCache[outputSampleID].t = vector.T
+			d.outputCache[outputSampleID].v = vector.Samples[i]
+		}
+
+		out := d.pool.GetStepVector(vector.T)
+		for outputSampleID, sample := range d.outputCache {
+			if sample.t == vector.T {
+				out.SampleIDs = append(out.SampleIDs, uint64(outputSampleID))
+				out.Samples = append(out.Samples, sample.v)
```

Comment:
Don't we need to `break` here? :thinking:

Reply:
Hm that would cause only one sample from the dedup cache to be added to the output vector. I wanted to avoid using a new dedup slice for each step vector, so this `t` field helps determine whether a sample from the current step vector should be added to the output. We set the `t` field [here](https://github.com/thanos-community/promql-engine/pull/151/files#diff-15c9fb08ea6d60219b7b77d429edda47df09c757ee3ba2262c84226f13b4cafcR65). If `t` does not equal to the current time for the output vector, we skip the sample because it's from a previous step.

I added a comment to help explain this.